login/login.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package login defines functionality used for allowing clients to authenticate
+// using USSO OAuth.
+package login
+
+import (
+	"encoding/json"
+	"io/ioutil"
+
+	"github.com/juju/usso"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/environschema.v1"
+	"gopkg.in/juju/environschema.v1/form"
+)
+
+type tokenGetter interface {
+	GetTokenWithOTP(username, password, otp, tokenName string) (*usso.SSOData, error)
+}
+
+// This is defined here to allow it to be stubbed out in tests
+var ussoServer tokenGetter = usso.ProductionUbuntuSSOServer
+
+// loginUSSO reads login parameters from ctxt.Stdin and then returns an
+// oauth token from Ubuntu SSO.
+func loginUSSO(filler form.Filler, store TokenStore) (*usso.SSOData, error) {
+	login, err := filler.Fill(ussoLoginForm)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot read login parameters")
+	}
+	tok, err := ussoServer.GetTokenWithOTP(
+		login["username"].(string),
+		login["password"].(string),
+		login["otp"].(string),
+		"charm",
+	)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot get token")
+	}
+	if err := store.Put(tok); err != nil {
+		return nil, errgo.Notef(err, "cannot save token")
+	}
+	return tok, nil
+}
+
+var ussoLoginForm = form.Form{
+	Fields: environschema.Fields{
+		"email": environschema.Attr{
```

Comment:
These might not look good when used with the IOFiller be prepared to change the field names, and descriptions.